src/data/clean_data.py

```diff
@@ -0,0 +1,245 @@
+    column_list = df.columns.values.tolist()
+    return column_list
+
+
+class DataCleaner:
+    """
+    Instantiated with a dataframe its column names, and size of lowest_similarity and ngram_size. 
+    DataCleaner methods to provide a smart deduplication of column values in a dataframe.
+    Utilizes a form of Term Frequency, Inverse Document Frequency to determine similarity between values.
+
+
+    Parameters:
+        path: Pandas.DataFrame.
+            Use _csv_to_df to provide df from csv.
+        
+        lowest_similarity: float
+            Lowest similarity percentage of column values to combine.
+                i.e. 0.9 = words with 90% or more similarity are combined.
+
+        column_name: str
+            A column name from path. Use _get_df_columns to get a list of columns.
+
+        ngram_size: int
+            Size of string chunks used to assess similarity between two values.
+            3 is normally best but values between 2 and 5 can work.
+
+    Returns:
+        A Pandas DataFrame with similar values of column_name combined.
+
+    """
+
+    def __init__(self, path: pd.DataFrame, lowest_similarity: float, column_name: str, ngram_size: int):
+        self.df = path
+        self.lowest_similarity = lowest_similarity
+        self.column_name = column_name
+        self.ngram_size = ngram_size
+        
+
+    def _ngrams(self, string):
+        """
+        Takes input `string`, removes unwanted chars and returns a list of slices of input `string`.
+
+        Parameters:
+            String: A string to be chunked into size of self.ngram_size.
+
+        Returns:
+            A list of strings. 
+                i.e. testing with self.ngram_size of 3
+                    [' Te', 'Tes', 'est', 'sti', 'tin', 'ing', 'ng ']
+        """
+        string = str(string)
+        string = fix_text(string) # fix text
+        string = string.encode("ascii", errors="ignore").decode() #remove non ascii chars
+        string = string.lower()
+        chars_to_remove = [")","(",".","|","[","]","{","}","'"]
+        rx = '[' + re.escape(''.join(chars_to_remove)) + ']'
+        string = re.sub(rx, '', string)
+        string = string.replace('&', 'and')
+        string = string.replace(',', ' ')
+        string = string.replace('-', ' ')
+        string = string.title() # normalise case - capital at start of each word
+        string = re.sub(' +',' ',string).strip() # get rid of multiple spaces and replace with a single
+        string = ' '+ string +' ' # pad names for ngrams...
+        string = re.sub(r'[,-./]|\sBD',r'', string)
+        ngrams = zip(*[string[i:] for i in range(self.ngram_size)])
+        return [''.join(ngram) for ngram in ngrams]
+
+
+    def _awesome_cossim_top(self, A, B, ntop):
+        """
+        True magic, an improvment on scikit-learn's cosine_similarity function.
+        Thanks to ING BANK.
+
+        ING definition:
+            This function will return a matrxi C in CSR format, where
```

Comment:
matrxi -> matrix

Reply:
You would think a 30+ billion dollar company would spell check so I didn't have to lol 😂